app/models/assignment.rb

```diff
@@ -224,7 +224,41 @@ def latest_due_date
     due_dates.compact.max
   end
 
-  def past_collection_date?(section=nil)
+  def all_grouping_collection_dates
+    submission_rule_max_hours = Assignment.where(id: self)
+                                          .joins(:groupings)
+                                          .joins(submission_rule: :periods)
+                                          .reorder(nil)
+                                          .group('groupings.id')
+                                          .sum(:hours)
+    all_grouping_due_dates.map do |grouping_id, due_date|
+      max_hours = submission_rule_max_hours[grouping_id] || 0
+      [grouping_id, due_date + max_hours.hours]
+    end.to_h
+  end
+
+  def all_grouping_due_dates
+    section_due_dates = Assignment.where(id: self)
+                                  .joins(groupings: [inviter: :section])
```

Comment:
You actually need the explicit `where` clause or else the `section_due_date` is just the first `section_due_date` that belongs to that `section` regardless of whether it is for this assignment or another assignment that the inviter has a grouping for. 

I will clean up the query a bit because it is overly complicated.